File: src/com/minecraft/moonlake/api/region/RegionVector2D.java
/*
 * Copyright (C) 2017 The MoonLake Authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package com.minecraft.moonlake.api.region;

import com.minecraft.moonlake.api.VectorConvertible;
import com.minecraft.moonlake.api.player.depend.WorldEditVector;
import com.minecraft.moonlake.util.StringUtil;
import com.minecraft.moonlake.validate.Validate;
import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.configuration.serialization.ConfigurationSerialization;
import org.bukkit.util.Vector;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * <h1>RegionVector2D</h1>
 * 区域矢量 2D（详细doc待补充...）
 *
 * @version 1.0
 * @author Month_Light
 */
public class RegionVector2D implements VectorConvertible, ConfigurationSerializable, Comparable<RegionVector2D> {

    /**
     * 区域矢量 2D ZERO 值
     */
    public final static RegionVector2D ZERO = new RegionVector2D(0d, 0d);

    static {

        ConfigurationSerialization.registerClass(RegionVector2D.class);
    }

    protected final double x;
    protected final double z;

    /**
     * 区域矢量 2D 构造函数
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D(double x, double z) {

        this.x = x;
        this.z = z;
    }

    /**
     * 区域矢量 2D 构造函数
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D(float x, float z) {

        this.x = x;
        this.z = z;
    }

    /**
     * 区域矢量 2D 构造函数
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D(int x, int z) {

        this.x = x;
        this.z = z;
    }

    /**
     * 区域矢量 2D 构造函数
     *
     * @param other 区域矢量 2D
     * @throws IllegalArgumentException 如果区域矢量 2D 对象为 {@code null} 则抛出异常
     */
    public RegionVector2D(RegionVector2D other) {

        Validate.notNull(other, "The other vector object is null.");

        this.x = other.x;
        this.z = other.z;
    }

    /**
     * 区域矢量 2D 构造函数
     */
    public RegionVector2D() {

        this.x = 0d;
        this.z = 0d;
    }

    /**
     * 获取此区域矢量 2D 的 X
     *
     * @return X
     */
    public double getX() {

        return x;
    }

    /**
     * 获取此区域矢量 2D 的 Z
     *
     * @return Z
     */
    public double getZ() {

        return z;
    }

    /**
     * 获取此区域矢量 2D 的方块 X
     *
     * @return 方块 X
     */
    public int getBlockX() {

        return (int) Math.round(x);
    }

    /**
     * 获取此区域矢量 2D 的方块 Z
     *
     * @return 方块 Z
     */
    public int getBlockZ() {

        return (int) Math.round(z);
    }

    /**
     * 设置此区域矢量 2D 的 X
     *
     * @param x X
     */
    public RegionVector2D setX(double x) {

        return new RegionVector2D(x, z);
    }

    /**
     * 设置此区域矢量 2D 的 Z
     *
     * @param z Z
     */
    public RegionVector2D setZ(double z) {

        return new RegionVector2D(x, z);
    }

    /**
     * 设置此区域矢量 2D 的 X
     *
     * @param x X
     */
    public RegionVector2D setX(int x) {

        return new RegionVector2D(x, z);
    }

    /**
     * 设置此区域矢量 2D 的 Z
     *
     * @param z Z
     */
    public RegionVector2D setZ(int z) {

        return new RegionVector2D(x, z);
    }

    /**
     * 将指定区域矢量 2D 和此区域矢量 2D 相加
     *
     * @param vector 区域矢量 2D
     */
    public RegionVector2D add(RegionVector2D vector) {

        return new RegionVector2D(x + vector.x, z + vector.z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相加
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D add(double x, double z) {

        return new RegionVector2D(this.x + x, this.z + z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相加
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D add(int x, int z) {

        return new RegionVector2D(this.x + x,  this.z + z);
    }

    /**
     * 将指定区域矢量 2D 和此区域矢量 2D 相减
     *
     * @param vector 区域矢量 2D
     */
    public RegionVector2D subtract(RegionVector2D vector) {

        return new RegionVector2D(x - vector.x, z - vector.z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相减
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D subtract(double x, double z) {

        return new RegionVector2D(this.x - x, this.z - z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相减
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D subtract(int x, int z) {

        return new RegionVector2D(this.x - x, this.z - z);
    }

    /**
     * 将指定区域矢量 2D 和此区域矢量 2D 相乘
     *
     * @param vector 区域矢量 2D
     */
    public RegionVector2D multiply(RegionVector2D vector) {

        return new RegionVector2D(x * vector.x, z * vector.z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相乘
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D multiply(double x, double z) {

        return new RegionVector2D(this.x * x, this.z * z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相乘
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D multiply(int x, int z) {

        return new RegionVector2D(this.x * x, this.z * z);
    }

    /**
     * 将指定 N 和此区域矢量 2D 相乘
     *
     * @param n N
     */
    public RegionVector2D multiply(double n) {

        return new RegionVector2D(this.x * n, this.z * n);
    }

    /**
     * 将指定 N 和此区域矢量 2D 相乘
     *
     * @param n N
     */
    public RegionVector2D multiply(float n) {

        return new RegionVector2D(this.x * n, this.z * n);
    }

    /**
     * 将指定 N 和此区域矢量 2D 相乘
     *
     * @param n N
     */
    public RegionVector2D multiply(int n) {

        return new RegionVector2D(this.x * n, this.z * n);
    }

    /**
     * 将指定区域矢量 2D 和此区域矢量 2D 相除
     *
     * @param vector 区域矢量 2D
     */
    public RegionVector2D divide(RegionVector2D vector) {

        return new RegionVector2D(x / vector.x, z / vector.z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相除
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D divide(double x, double z) {

        return new RegionVector2D(this.x / x, this.z / z);
    }

    /**
     * 将指定 X Z 和此区域矢量 2D 相除
     *
     * @param x X
     * @param z Z
     */
    public RegionVector2D divide(int x, int z) {

        return new RegionVector2D(this.x / x, this.z / z);
    }

    /**
     * 将指定 N 和此区域矢量 2D 相除
     *
     * @param n N
     */
    public RegionVector2D divide(double n) {

        return new RegionVector2D(this.x / n, this.z / n);
    }

    /**
     * 将指定 N 和此区域矢量 2D 相除
     *
     * @param n N
     */
    public RegionVector2D divide(float n) {

        return new RegionVector2D(this.x / n, this.z / n);
    }

    /**
     * 将指定 N 和此区域矢量 2D 相除
     *
     * @param n N
     */
    public RegionVector2D divide(int n) {

        return new RegionVector2D(this.x / n, this.z / n);
    }

    /**
     * 获取此区域矢量 2D 的长度
     *
     * @return 长度
     */
    public double length() {

        return Math.sqrt(lengthSq());
    }

    /**
     * 获取此区域矢量 2D 的长度平方
     *
     * @return 长度平方
     */
    public double lengthSq() {

        return x * x + z * z;
    }

    /**
     * 获取指定区域矢量 2D 和此区域矢量 2D 的距离
     *
     * @param vector 区域矢量 2D
     * @return 距离
     */
    public double distance(RegionVector2D vector) {

        return Math.sqrt(distanceSq(vector));
    }

    /**
     * 获取指定区域矢量 2D 和此区域矢量 2D 的距离平方
     *
     * @param vector 区域矢量 2D
     * @return 距离平方
     */
    public double distanceSq(RegionVector2D vector) {

        return Math.pow(x - vector.x, 2d) + Math.pow(z - vector.z, 2d);
    }

    /**
     * 获取此区域矢量 2D 的归一化
     *
     * @return 使归一化的区域矢量 2D
     */
    public RegionVector2D normalize() {

        return divide(length());
    }

    /**
     * 将此区域矢量 2D 转换为区域矢量
     *
     * @return 区域矢量
     */
    public RegionVector toRegionVector() {

        return toRegionVector(0d);
    }

    /**
     * 将此区域矢量 2D 转换为区域矢量
     *
     * @param y Y
     * @return 区域矢量
     */
    public RegionVector toRegionVector(double y) {

        return new RegionVector(x, y, z);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        RegionVector2D that = (RegionVector2D) o;

        if (Double.compare(that.x, x) != 0) return false;
        return Double.compare(that.z, z) == 0;
    }

    @Override
    public int hashCode() {
        int result;
        long temp;
        temp = Double.doubleToLongBits(x);
        result = (int) (temp ^ (temp >>> 32));
        temp = Double.doubleToLongBits(z);
        result = 31 * result + (int) (temp ^ (temp >>> 32));
        return result;
    }

    @Override
    public String toString() {
        return "RegionVector2D{" +
                "x=" + x +
                ", z=" + z +
                '}';
    }

    @Override
    public int compareTo(RegionVector2D o) {

        if(x != o.x)
            return Double.compare(x, o.x);
        if(z != o.z)
            return Double.compare(z, o.z);
        return 0;
    }

    @Override
    public Map<String, Object> serialize() {

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("x", x);
        result.put("z", z);
        return result;
    }

    public static RegionVector2D deserialize(Map<String, Object> args) {

        double x = StringUtil.parseDouble(args.get("x"));
        double z = StringUtil.parseDouble(args.get("z"));
        return new RegionVector2D(x, z);
    }

    @Override
    public Vector asBukkitVector() {

        return new Vector(x, 0d, z);
    }

    @Override
    public WorldEditVector asWorldEditVector() {

        return new WorldEditVector(x, 0d, z);
    }

    @Override
    public RegionVector asRegionVector() {

        return new RegionVector(x, 0d, z);
    }

    @Override
    public RegionBlockVector asRegionBlockVector() {

        return new RegionBlockVector(x, 0d, z);
    }
}

File: API/src/main/kotlin/com/minecraft/moonlake/api/region/RegionVector2D.kt
/*
 * Copyright (C) 2016-2017 The MoonLake (mcmoonlake@hotmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.minecraft.moonlake.api.region

import com.minecraft.moonlake.api.util.ComparisonChain
import com.minecraft.moonlake.api.parseDouble
import org.bukkit.Location
import org.bukkit.World
import org.bukkit.configuration.serialization.ConfigurationSerializable

open class RegionVector2D(
        val x: Double,
        val z: Double) : ConfigurationSerializable, Comparable<RegionVector2D> {

    /** constructor */

    constructor(x: Float, z: Float) : this(x.toDouble(), z.toDouble())
    constructor(x: Int, z: Int) : this(x.toDouble(), z.toDouble())
    constructor(other: RegionVector2D) : this(other.x, other.z)
    constructor() : this(.0, .0)

    /** api */

    val blockX: Int
        get() = Math.round(x).toInt()

    val blockZ: Int
        get() = Math.round(z).toInt()

    fun setX(x: Double): RegionVector2D
            = RegionVector2D(x, z)

    fun setZ(z: Double): RegionVector2D
            = RegionVector2D(x, z)

    fun setX(x: Int): RegionVector2D
            = RegionVector2D(x.toDouble(), z)

    fun setZ(z: Int): RegionVector2D
            = RegionVector2D(x, z.toDouble())

    fun length(): Double
            = Math.sqrt(lengthSq())

    fun lengthSq(): Double
            = x * x + z * z

    fun distance(vector: RegionVector2D): Double
            = Math.sqrt(distanceSq(vector))

    fun distanceSq(vector: RegionVector2D): Double
            = Math.pow(x - vector.x, 2.0) + Math.pow(z - vector.z, 2.0)

    fun normalize(): RegionVector2D
            = div(length())

    fun toRegionVector(y: Double = .0): RegionVector
            = RegionVector(x, y, z)

    fun toRegionVector(y: Int = 0): RegionVector
            = RegionVector(x, y.toDouble(), z)

    fun toLocation(world: World, y: Double = .0, yaw: Float = .0f, pitch: Float = .0f): Location
            = Location(world, x, y, z, yaw, pitch)

    /** operator */

    operator fun inc(): RegionVector2D
            = RegionVector2D(x + 1.0, z + 1.0)

    operator fun dec(): RegionVector2D
            = RegionVector2D(x - 1.0, z - 1.0)

    operator fun plus(other: RegionVector2D): RegionVector2D
            = RegionVector2D(x + other.x, z + other.z)

    fun plus(x: Double, z: Double): RegionVector2D
            = RegionVector2D(this.x + x, this.z + z)

    fun plus(x: Int, z: Int): RegionVector2D
            = RegionVector2D(this.x + x, this.z + z)

    operator fun minus(other: RegionVector2D): RegionVector2D
            = RegionVector2D(x - other.x, z - other.z)

    fun minus(x: Double, z: Double): RegionVector2D
            = RegionVector2D(this.x - x, this.z - z)

    fun minus(x: Int, z: Int): RegionVector2D
            = RegionVector2D(this.x - x, this.z - z)

    operator fun times(other: RegionVector2D): RegionVector2D
            = RegionVector2D(x * other.x, z * other.z)

    operator fun times(other: Double): RegionVector2D
            = RegionVector2D(x * other, z * other)

    operator fun times(other: Float): RegionVector2D
            = RegionVector2D(x * other, z * other)

    operator fun times(other: Int): RegionVector2D
            = RegionVector2D(x * other, z * other)

    fun times(x: Double, z: Double): RegionVector2D
            = RegionVector2D(this.x * x, this.z * z)

    fun times(x: Int, z: Int): RegionVector2D
            = RegionVector2D(this.x * x, this.z * z)

    operator fun div(other: RegionVector2D): RegionVector2D
            = RegionVector2D(x / other.x, z / other.z)

    operator fun div(other: Double): RegionVector2D
            = RegionVector2D(x / other, z / other)

    operator fun div(other: Float): RegionVector2D
            = RegionVector2D(x / other, z / other)

    operator fun div(other: Int): RegionVector2D
            = RegionVector2D(x / other, z / other)

    fun div(x: Double, z: Double): RegionVector2D
            = RegionVector2D(this.x / x, this.z / z)

    fun div(x: Int, z: Int): RegionVector2D
            = RegionVector2D(this.x / x, this.z / z)

    /** significant */

    override fun equals(other: Any?): Boolean {
        if(other === this)
            return true
        if(other is RegionVector2D)
            return x == other.x && z == other.z
        return false
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + z.hashCode()
        return result
    }

    override fun toString(): String {
        return "RegionVector2D(x=$x, z=$z)"
    }

    override fun compareTo(other: RegionVector2D): Int {
        return ComparisonChain.start()
                .compare(x, other.x)
                .compare(z, other.z)
                .result
    }

    override fun serialize(): MutableMap<String, Any> {
        val result = LinkedHashMap<String, Any>()
        result.put("x", x)
        result.put("z", z)
        return result
    }

    /** static */

    companion object {

        @JvmField val ZERO = RegionVector2D(.0, .0)

        @JvmStatic
        @JvmName("deserialize")
        fun deserialize(args: Map<String, Any>): RegionVector2D {
            val x = args["x"]?.parseDouble() ?: .0
            val z = args["z"]?.parseDouble() ?: .0
            return RegionVector2D(x, z)
        }
    }
}
